### PR TITLE
[feature/chucker-gql] GQL support against latest develop 

### PR DESCRIPTION
**Pull request by alkber**
There was already an old implementation done [here ](https://github.com/ChuckerTeam/chucker/pull/70). This PR seems to be old one (2019) and way out of sync with develop branch. I have refactored it to meet the latest changes. 

This almost getting ready and not the final PR yet, need to get valuable input on the code quality and optimization if required. 

Sample Activity has also been provided for the demo

**Known Issues**  
Couldn't successfully run the sample until i commented this part of the code 

```
//        StrictMode.setVmPolicy(
//            StrictMode.VmPolicy.Builder()
//                .detectLeakedClosableObjects()
//                .penaltyLog()
//                .penaltyDeath()
//                .build()
//        )
//
//        StrictMode.setThreadPolicy(
//            StrictMode.ThreadPolicy.Builder()
//                .detectDiskReads()
//                .detectDiskWrites()
//                .penaltyLog()
//                .penaltyDeath()
//                .build()
//        )
```

### CODE REVIEW

File: sample/src/main/kotlin/com/chuckerteam/chucker/sample/GqlHttpTask.kt [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import retrofit2.converter.gson.GsonConverterFactory
+import retrofit2.http.Body
+import retrofit2.http.Headers
+import retrofit2.http.POST
+
+class GqlHttpTask(
+    client: OkHttpClient,
+) : HttpTask {
+
+    private val api = Retrofit.Builder()
```
Comment (MiSikora) [inline sample/src/main/kotlin/com/chuckerteam/chucker/sample/GqlHttpTask.kt:14]: I think that the vast majority of people will use GraphQL with something like Apollo. I would prefer to see a real-life scenario integration within our sample.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/support/RequestProcessor.kt [hunk 2]
```diff
@@ -75,4 +82,15 @@ internal class RequestProcessor(
                 null
             }
         }.firstOrNull()
+
+    private fun inferGqlRequest(body: String): String? {
```
Comment (MiSikora) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/support/RequestProcessor.kt:86]: Inferring GraphQL from a body won't work for GET requests as there is only URL query parameter for those.

It's a bit hard to have a full-proof solution as anyone could have query parameters or bodies that conform to GraphQL format. I have an idea of how to deal with it but let's just first implement handling of GET requests as well.

File: library/src/main/kotlin/com/chuckerteam/chucker/internal/data/entity/HttpTransaction.kt [hunk 3]
```diff
@@ -134,10 +136,13 @@ internal class HttpTransaction(
             return when (status) {
                 Status.Failed -> " ! ! !  $method $path"
                 Status.Requested -> " . . .  $method $path"
-                else -> "$responseCode $method $path"
+                else -> "$responseCode $method $path ${isGraphqlRequest.takeIf { true } ?: ""}"
             }
         }
 
+    val isGraphqlRequest: Boolean
+        get() = !gqlOperationName.isNullOrEmpty()
```
Comment (MiSikora) [inline library/src/main/kotlin/com/chuckerteam/chucker/internal/data/entity/HttpTransaction.kt:144]: This doesn't seem right. The `operationName` is an optional parameter in [GraphQL](https://graphql.org/learn/serving-over-http/#post-request). I think we can rely solely on `query`. Plus there is an additional issue for GET  requests, which I mentioned here.

**Review by MiSikora — CHANGES_REQUESTED** [review]
For a PR that introduces integration with a new specification, there are no tests. Please cover necessary scenarios such as GET and POST requests with `ChuckerInterceptor` as both application and network interceptor.

It would be also helpful if you could follow our PR template. Screenshots especially would be appreciated.

**Comment by MiSikora** [comment]
> **Known Issues** Couldn't successfully run the sample until i commented this part of the code
> 
> ```
> //        StrictMode.setVmPolicy(
> //            StrictMode.VmPolicy.Builder()
> //                .detectLeakedClosableObjects()
> //                .penaltyLog()
> //                .penaltyDeath()
> //                .build()
> //        )
> //
> //        StrictMode.setThreadPolicy(
> //            StrictMode.ThreadPolicy.Builder()
> //                .detectDiskReads()
> //                .detectDiskWrites()
> //                .penaltyLog()
> //                .penaltyDeath()
> //                .build()
> //        )
> ```

I experienced it in the past as well but I thought that #750 fixed it.

**Comment by cortinico** [comment]
Sorry for the late reply here.
I've noticed we do have another similar PR here: #805
I'd like to understand which of the two we prefer to move forward.

**Comment by alkber** [comment]
Please close this PR in respect of a better implementation at [https://github.com/ChuckerTeam/chucker/pull/805](url)

**Comment by cortinico** [comment]
> Please close this PR in respect of a better implementation at [https://github.com/ChuckerTeam/chucker/pull/805](url)

👍 We'll keep the disucssion going on #805

**Pull request closed without merging**